File: leetcode/editor/cn/202_快乐数_test.py
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        is_happy = self.solution.isHappy(19)
        self.assertTrue(is_happy)


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def isHappy(self, n: int) -> bool:
        ans: bool = False

        def next_n(a_num: int) -> int:
            result: int = 0
            while a_num:
                a_num, mod = divmod(a_num, 10)
                result += mod * mod
            return result

        slow, fast = next_n(n), next_n(next_n(n)),
        while True:
            fn = next_n(fast)
            if slow == 1 or fast == 1 or fn == 1:
                ans = True
                break
            if slow == fast:
                break
            slow, fast = next_n(slow), next_n(fn)
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()

File: leetcode/editor/cn/202_快乐数_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when202_thenSuccess(t *testing.T) {
	assert.Equal(t, true, isHappy(19))
	assert.Equal(t, false, isHappy(2))
}

//leetcode submit region begin(Prohibit modification and deletion)
func isHappy(n int) bool {
	// 不快乐数唯一循环序列 4 16 37 58 89 145 42 20
	// 若 n 的不快乐数是上述序列中的任意一个值，n 便是不快乐数
	ans := false
	for true {
		n = nextN(n)
		if n == 1 {
			ans = true
			break
		} else if n == 4 {
			break
		}
	}
	return ans

}
func nextN(n int) int {
	ans := 0
	for n != 0 {
		mod := n % 10
		ans += mod * mod
		n /= 10
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)
